// @kkbox/kkbox-js-sdk/dist/api/AlbumFetcher.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _Endpoint = require("../Endpoint");

var _Fetcher2 = _interopRequireDefault(require("./Fetcher"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

/**
 * Fetch metadata and tracks of a album.
 * @see https://docs-en.kkbox.codes/v1.1/reference#albums
 */
var AlbumFetcher =
/*#__PURE__*/
function (_Fetcher) {
  _inherits(AlbumFetcher, _Fetcher);

  /**
   * @ignore
   */
  function AlbumFetcher(http) {
    var _this;

    var territory = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'TW';

    _classCallCheck(this, AlbumFetcher);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(AlbumFetcher).call(this, http, territory));
    /**
     * @ignore
     */

    _this.albumID = undefined;
    return _this;
  }
  /**
   * Set the album fetcher.
   *
   * @param {string} albumID - The ID of an album.
   * @return {AlbumFetcher}
   * @see https://docs-en.kkbox.codes/v1.1/reference#albums-album_id
   */


  _createClass(AlbumFetcher, [{
    key: "setAlbumID",
    value: function setAlbumID(albumID) {
      this.albumID = albumID;
      return this;
    }
    /**
     * Fetch metadata of the album you create.
     *
     * @return {Promise}
     * @example api.albumFetcher.setAlbumID('KmRKnW5qmUrTnGRuxF').fetchMetadata();
     * @see https://docs-en.kkbox.codes/v1.1/reference#albums-album_id
     */

  }, {
    key: "fetchMetadata",
    value: function fetchMetadata() {
      return this.http.get(_Endpoint.ALBUMS + '/' + this.albumID, {
        territory: this.territory
      });
    }
    /**
     * Get KKBOX web widget uri of the album.
     * @example https://widget.kkbox.com/v1/?id=4qtXcj31wYJTRZbb23&type=album
     * @return {string}
     */

  }, {
    key: "getWidgetUri",
    value: function getWidgetUri() {
      return "https://widget.kkbox.com/v1/?id=".concat(this.albumID, "&type=album");
    }
    /**
     * Get tracks in an album. Result will be return with pagination.
     *
     * @param {number} [limit] - The size for one page.
     * @param {number} [offset] - The offset index for first element.
     * @return {Promise}
     * @example api.albumFetcher.setAlbumID('KmRKnW5qmUrTnGRuxF').fetchTracks();
     * @see https://docs-en.kkbox.codes/v1.1/reference#albums-album_id-tracks
     */

  }, {
    key: "fetchTracks",
    value: function fetchTracks() {
      var limit = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : undefined;
      var offset = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : undefined;
      return this.http.get(_Endpoint.ALBUMS + '/' + this.albumID + '/tracks', {
        territory: this.territory,
        limit: limit,
        offset: offset
      });
    }
  }]);

  return AlbumFetcher;
}(_Fetcher2["default"]);

exports["default"] = AlbumFetcher;
