// @kkbox/kkbox-js-sdk/dist/api/MoodStationFetcher.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _Endpoint = require("../Endpoint");

var _Fetcher2 = _interopRequireDefault(require("./Fetcher"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

/**
 * Get mood stations.
 * @see https://docs-en.kkbox.codes/v1.1/reference#mood-stations
 */
var MoodStationFetcher =
/*#__PURE__*/
function (_Fetcher) {
  _inherits(MoodStationFetcher, _Fetcher);

  /**
   * @ignore
   */
  function MoodStationFetcher(http) {
    var _this;

    var territory = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'TW';

    _classCallCheck(this, MoodStationFetcher);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(MoodStationFetcher).call(this, http, territory));
    /**
     * @ignore
     */

    _this.moodStationID = undefined;
    return _this;
  }
  /**
   * Fetch all mood stations.
   *
   * @return {Promise}
   * @example api.moodStationFetcher.fetchAllMoodStations();
   * @see https://docs-en.kkbox.codes/v1.1/reference#moodstations
   */


  _createClass(MoodStationFetcher, [{
    key: "fetchAllMoodStations",
    value: function fetchAllMoodStations() {
      return this.http.get(_Endpoint.MOOD_STATIONS, {
        territory: this.territory
      });
    }
    /**
     * Init the mood station fetcher.
     *
     * @param {string} moodStationID - The ID of a mood station.
     * @param {string} [territory = 'TW'] - ['TW', 'HK', 'SG', 'MY', 'JP'] The territory of a mood station.
     * @return {MoodStation}
     * @see https://docs-en.kkbox.codes/v1.1/reference#moodstations-station_id
     */

  }, {
    key: "setMoodStationID",
    value: function setMoodStationID(moodStationID) {
      var territory = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'TW';
      this.moodStationID = moodStationID;
      this.territory = territory;
      return this;
    }
    /**
     * Fetch the mood station's metadata.
     *
     * @return {Promise}
     * @example api.moodStationFetcher.setMoodStationID('StGZp2ToWq92diPHS7').fetchMetadata();
     * @see https://docs-en.kkbox.codes/v1.1/reference#moodstations-station_id
     */

  }, {
    key: "fetchMetadata",
    value: function fetchMetadata() {
      return this.http.get(_Endpoint.MOOD_STATIONS + '/' + this.moodStationID, {
        territory: this.territory
      });
    }
  }]);

  return MoodStationFetcher;
}(_Fetcher2["default"]);

exports["default"] = MoodStationFetcher;
